File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.cpp
class Solution {
private:
    int findMaxSizeSubarraySum(vector<int>& nums, int target){
        if(target < 0)
            return -1;

        int n = nums.size();

        int i = 0, j = 0;
        int currSum = 0;
        int maxSize = -1;

        while(j<n){
            currSum += nums[j];
            
            while(i <= j && currSum > target){
                currSum -= nums[i++];
            }

            if(currSum == target)
                    maxSize = max(maxSize, (j-i+1));

            j++;
        }

        return maxSize;
    }
public:
    int minOperations(vector<int>& nums, int x) {
        int n = nums.size();
        int sum = accumulate(nums.begin(), nums.end(), 0);
        int maxSize = findMaxSizeSubarraySum(nums, sum - x);
        return maxSize == -1 ? -1 : (n - maxSize);
    }
};

File: 1658-minimum-operations-to-reduce-x-to-zero/1658-minimum-operations-to-reduce-x-to-zero.java
class Solution {
    private int maxSizeSubarray(int[] nums, int sum){
        int n = nums.length;
        int maxLen = -1;
        int i = 0, j = 0;
        int currSum = 0;
        while(j < n){
            currSum += nums[j];
            while(i <=j && currSum > sum){
                currSum -= nums[i++];
            }
            if(currSum == sum){
                maxLen = Math.max(maxLen, (j - i + 1));
            }
            j++;
        }
        return maxLen;
    }
    
    public int minOperations(int[] nums, int x) {
        int n = nums.length;
        int totalSum = 0;
        for(int num : nums){
            totalSum += num;
        }
        int maxLen = maxSizeSubarray(nums, totalSum - x);
        if(maxLen == -1){
            return -1;
        }
        return n - maxLen;
    }
}